### PR TITLE
Projectile generator datums and projectile component signals

### PR DESCRIPTION
**Pull request by silicons**
A little something to get myself back in coding, that will hopefully be useful later in reusing special effects like repulsion waves/explosions/stuff from projectiles. 
Gyrojets converted to the new system, everything else is still on the "normal" code. The new system defaults to off/not used. 
The generator itself is by default, put on the ammo casing rather than the gun or projectile. This makes it not need to be remade everytime something is fired for energy weapons, although it will still have a performance/memory cost hit if this is used for ballistics. Currently don't know how to do this for ballistics without having to make one for each casing (as guns can use more than one type of ammo, so can't put it on the gun.)
Components are put on generators or projectiles
Also someone please, please look at how I'm handling GC for projectile generators, and make sure I'm not making a memory leak...

:cl:
code: Projectile generator datums are added. 
/:cl:

### CODE REVIEW

File: code/datums/components/projectile_effect/_projectile_effect.dm [hunk 1]
```diff
@@ -0,0 +1,9 @@
+//No, this doesn't go on /obj/item/projectile, this goes on /datum/projectile_generator.
+/datum/component/projectile_effect
```
Comment (ninjanomnom) [inline code/datums/components/projectile_effect/_projectile_effect.dm]: stop

This should not exist and has no purpose

**Comment by optimumtact** [comment]
I don't see the point of this, projectiles are already highly abstracted due to cartridges bullets and projectile types.

These behaviours on most projectiles don't really make sense outside of projectile system either, it's like componetisation for the sake of it.

File: code/datums/components/projectile_effect/onhit/_onhit.dm [hunk 2]
```diff
@@ -0,0 +1,8 @@
+/datum/component/projectile_effect/onhit
```
Comment (ninjanomnom) [inline code/datums/components/projectile_effect/onhit/_onhit.dm]: what the fuck

File: code/datums/components/projectile_effect/onhit/explosive.dm [hunk 3]
```diff
@@ -0,0 +1,36 @@
+/datum/component/projectile_effect/onhit/explosive
```
Comment (ninjanomnom) [inline code/datums/components/projectile_effect/onhit/explosive.dm]: make a regular component for things which explode on hitting stuff

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by kevinz000** [comment]
eh yeah this isn't really needed

**Pull request closed without merging**